model/infinispan/src/main/java/org/keycloak/models/cache/infinispan/authorization/entities/CachedPolicy.java

```diff
@@ -38,27 +40,91 @@
     private String type;
     private DecisionStrategy decisionStrategy;
     private Logic logic;
-    private Map<String, String> config;
     private String name;
     private String description;
     private String resourceServerId;
-    private Set<String> associatedPoliciesIds;
-    private Set<String> resourcesIds;
-    private Set<String> scopesIds;
+    private Function<Supplier<Policy>, Set<String>> associatedPoliciesIds;
+    private Function<Supplier<Policy>, Set<String>> resourcesIds;
+    private Function<Supplier<Policy>, Set<String>> scopesIds;
+    private Function<Supplier<Policy>, Map<String, String>> config;
     private final String owner;
 
     public CachedPolicy(Long revision, Policy policy) {
         super(revision, policy.getId());
         this.type = policy.getType();
         this.decisionStrategy = policy.getDecisionStrategy();
         this.logic = policy.getLogic();
-        this.config = new HashMap(policy.getConfig());
         this.name = policy.getName();
         this.description = policy.getDescription();
         this.resourceServerId = policy.getResourceServer().getId();
-        this.associatedPoliciesIds = policy.getAssociatedPolicies().stream().map(Policy::getId).collect(Collectors.toSet());
-        this.resourcesIds = policy.getResources().stream().map(Resource::getId).collect(Collectors.toSet());
-        this.scopesIds = policy.getScopes().stream().map(Scope::getId).collect(Collectors.toSet());
+
+        if (policy.isFetched("associatedPolicies")) {
+            Set<String> cached = policy.getAssociatedPolicies().stream().map(Policy::getId).collect(Collectors.toSet());
+            this.associatedPoliciesIds = supplier -> cached;
+        } else {
+            this.associatedPoliciesIds = new Function<Supplier<Policy>, Set<String>>() {
+                Set<String> cached;
```

Comment:
All the introduced `cached` fields in this class should be `private final AtomicReference`s. The update logic needs to count on possibility of `cached` being computed by two concurrent threads (see below).